### ISSUE TITLE
Projection in CH for ReplicatedMergeTree not working

### ISSUE THREAD

**Issue by priyanss1** [ClickHouse/ClickHouse#53147]
### I am adding projection in existing table but not sure if CH is returning result from projection .
Below are the queries .

**Create Table Statement**

`create table default.z_local on cluster '{cluster}' (Browser String, Country UInt8, F Float64,_ingest_time DateTime default now64(3,'UTC')
Engine= ReplicatedMergeTree('/clickhouse/{installation}/report/report/{shard}/{database}/zzz','{replica}')
order by Browser;

CREATE TABLE default.z ON CLUSTER '{cluster}' AS default.z_local
    ENGINE = Distributed('report', default, z_local, rand());

insert into default.z_local(Browser,Country,F) select toString(number%9999),number%33, 1
    from numbers(100000000);
`

**Alter table and added projection** 

`alter table default.z_local ON CLUSTER '{cluster}' add projection pp
   (select Browser,Country, count(), sum(F)
    group by Browser,Country);

alter table default.z_local ON CLUSTER '{cluster}' materialize projection pp;
`

**Query** 
`SELECT
    sum(F),
    Browser
FROM default.z
GROUP BY Browser ;
`

**Query Log**
`SELECT query, projections FROM clusterAllReplicas(report, system.query_log) WHERE query_id ='71ea9d9e-29a8-4ac1-9033-7c918c1d804d'`

I can see the query time reduced after adding projection but the result for above query  in  system.query_log should return projections columns with name of projection but I am getting empty value in projections column.

Please suggest

![image](https://github.com/ClickHouse/ClickHouse/assets/131259972/02474f77-3512-412e-8ec1-36b4ea9002b6)

![image](https://github.com/ClickHouse/ClickHouse/assets/131259972/9ba74bb8-4423-4480-bfa8-6137c5eaf51c)

```[tasklist]
### Tasks
```

**Comment by melvynator** [comment]
What version are you using?

**Comment by melvynator** [comment]
I have replicated your issue in `23.5.4.25` but it seems to be fixed on latest. Please use the EXPLAIN indexes=1 to ensure the projection is used correctly. 

```
SELECT version();

CREATE TABLE default.melvyn_test
(
    `sampled_at` DateTime,
    `playback_type` String,
    `country_code` Enum8('France' = 1, 'Allemagne' = 2, 'Belgique' = 3),
    `device` Enum8('Iphone' = 1, 'Android' = 2, 'Other' = 3),
    PROJECTION viewers_melvyn_2_country_code
    (
        SELECT *
        ORDER BY 
            country_code
    ),
    PROJECTION viewers_melvyn_2_device
    (
        SELECT *
        ORDER BY 
            device
    )
)
ENGINE = MergeTree()
ORDER BY toStartOfMinute(sampled_at); 

INSERT INTO melvyn_test SELECT * FROM generateRandom('`sampled_at` DateTime,`playback_type` String, `country_code` Enum8(\'France\' = 1, \'Allemagne\' = 2, \'Belgique\' = 3), `device` Enum8(\'Iphone\' = 1, \'Android\' = 2, \'Other\' = 3)', 1, 10, 2) LIMIT 1000000;

SELECT * FROM melvyn_test WHERE device='Iphone' LIMIT 1;

SELECT event_time, projections, query FROM clusterAllReplicas(default, system.query_log) WHERE query_id='id_query' ORDER BY event_time DESC;

EXPLAIN indexes=1 SELECT * FROM melvyn_test WHERE device='Iphone' LIMIT 1;

DROP TABLE melvyn_test;
```

Can you confirm that the query with EXPLAIN indexes=1 return something containing:

`ReadFromMergeTree (viewers_melvyn_2_device)`

**Comment by priyanss1** [comment]
> What version are you using?

23.4.6.25

**Comment by priyanss1** [comment]
> I have replicated your issue in `23.5.4.25` but it seems to be fixed on latest. Please use the EXPLAIN indexes=1 to ensure the projection is used correctly.
> 
> ```
> SELECT version();
> 
> CREATE TABLE default.melvyn_test
> (
>     `sampled_at` DateTime,
>     `playback_type` String,
>     `country_code` Enum8('France' = 1, 'Allemagne' = 2, 'Belgique' = 3),
>     `device` Enum8('Iphone' = 1, 'Android' = 2, 'Other' = 3),
>     PROJECTION viewers_melvyn_2_country_code
>     (
>         SELECT *
>         ORDER BY 
>             country_code
>     ),
>     PROJECTION viewers_melvyn_2_device
>     (
>         SELECT *
>         ORDER BY 
>             device
>     )
> )
> ENGINE = MergeTree()
> ORDER BY toStartOfMinute(sampled_at); 
> 
> INSERT INTO melvyn_test SELECT * FROM generateRandom('`sampled_at` DateTime,`playback_type` String, `country_code` Enum8(\'France\' = 1, \'Allemagne\' = 2, \'Belgique\' = 3), `device` Enum8(\'Iphone\' = 1, \'Android\' = 2, \'Other\' = 3)', 1, 10, 2) LIMIT 1000000;
> 
> SELECT * FROM melvyn_test WHERE device='Iphone' LIMIT 1;
> 
> SELECT event_time, projections, query FROM clusterAllReplicas(default, system.query_log) WHERE query_id='id_query' ORDER BY event_time DESC;
> 
> EXPLAIN indexes=1 SELECT * FROM melvyn_test WHERE device='Iphone' LIMIT 1;
> 
> DROP TABLE melvyn_test;
> ```
> 
> Can you confirm that the query with EXPLAIN indexes=1 return something containing:
> 
> `ReadFromMergeTree (viewers_melvyn_2_device)`


YES,I can see the explain query return ReadFromMergeTree (pp) -----pp is projection name for table I have created.

![image](https://github.com/ClickHouse/ClickHouse/assets/131259972/22509128-21b6-4b3c-aa6d-ff294203a50a)

**Comment by priyanss1** [comment]
Thank you @melvynator  for response  ,I will upgrade the version and test

**Comment by priyanss1** [comment]
Its showing projection after upgrade to latest .
Thank you @melvynator 

![image](https://github.com/ClickHouse/ClickHouse/assets/131259972/d24900d8-ea05-48f6-ab1b-a8e153e9dbe9)

**Comment by sunny19930321** [comment]
whether this fix was merged to the latest version 23.3
